Check and clean up the examples, and  add error chains (#505)
neo4j/docs-status-codes: modules/ROOT/pages/errors/gql-errors/22ND1.adoc

Try denying privileges to the role:

[source, cypher]
----
DENY MATCH {*} ON GRAPH secretDb NODES * TO otherrole, role;
----

The query returns a chain of errors where GQLSTATUS 22ND1 is the cause of GQLSTATUS xref:errors/gql-errors/42001.adoc[42001]:

.GQLSTATUS error chain
[source, error]
----
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'DENY MATCH {*} ON GRAPH secretDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
----

The error message contains a subset of the original query that caused the issue.
Since `otherrole` is not granted to an auth rule, it is not included.

== Possible solutions

Consider if it is possible to implement the security model without using denied privileges.

For example, you can replace the generic deny from the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:

> Grant match privileges to the role for specific graphs instead of denying access to all nodes in the `secretDb` graph:

Reviewer: feels slightly odd that both of these lines end with `:`?

[[example-scenario-22nd1]]
== Example scenario

The example assumes that you have a database called `secretDb` and you want to deny access to all nodes in that database for a role called `role`.

> You also have an auth rule called `authrule` that grants the role `role` to users who meet certain criteria.

Reviewer: ```suggestion
You also have two roles, called `role` and `otherrole`, and an auth rule, called `authrule`.
```
since the first thing you do is grant role to authrule and otherrole didn't have any mention but are used in the example

. Grant the role `role` to the auth rule:
+
[source, cypher]
----
CYPHER 25 GRANT ROLE role TO AUTH RULE authrule;

> ----

Reviewer: To avoid having the confusion around granting/revoking the role from the auth rule we could just have this in the setup sentence instead of example looking:

```
You also have two roles, called `role` and `otherrole`, and an auth rule, called `authrule`, where `role` is granted to `authrule` and `otherrole` isn't.
```

> . Try denying privileges from the role:

Reviewer: ```suggestion
. Try denying privileges to the role:
```

+
[source, cypher]
----
DENY MATCH {*} ON GRAPH secretDb NODES * TO otherrole, role;
----
+
The query returns a chain of errors where GQLSTATUS 22ND1 is the cause of GQLSTATUS xref:errors/gql-errors/42001.adoc[42001]:
+
.GQLSTATUS error chain
[source, error]
----
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'DENY MATCH {*} ON GRAPH secretDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
----
+
The error message contains a subset of the original query that caused the issue.
Since `otherrole` is not granted to an auth rule, it is not included.

== Possible solutions

Consider if it is possible to implement the security model without using denied privileges.

> For example, you can replace the generic grant from step 1 in the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:

Reviewer: ```suggestion
For example, you can replace the generic deny from step 2 in the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:
```

Author: Oh, I see. It was wrong in the first place.

. Revoke the role from the auth rule to reproduce the original scenario:
+
[source, cypher]
----

> REVOKE ROLE role FROM AUTH RULE authrule;

Reviewer: why are we revoking the role? isn't the point that we want the role on the auth rule and are now fixing so we can have the privilege (as the DENY version wasn't allowed)

Author: Because that's what is said on line 51.

Reviewer: > For example, you can replace the generic grant from step 1 in the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:

doesn't this just say to replace the `DENY MATCH {*} ON GRAPH secretDb NODES * TO otherrole, role;` with a new privilege? And nothing about the role/auth rule assignment...

The point is to show that this new GRANT privilege can be granted to a role assigned to an auth rule -> if you revoke the role from the auth rule then the DENY privilege would also be fine...

Author: It says to replace the grant, not the deny.

Reviewer: yeah, probably because we are bad at not calling both granting and denying a privilege a grant... (and since the original sentence didn't have a `step` part it could therefore have referred to either assignment, but since the error is about the DENY specifically I know that that is what "generic grant" refers to here... hence my other comments changing them to "generic deny" instead

Author: When run the alternative without revoking the role from the auth rule:
`GRANT MATCH {*} ON GRAPH publicDb, documentationDb NODES * TO role, otherrole;
`
I still get the error.

```
neo4j@system> DENY MATCH {*} ON GRAPH secretDb NODES * TO otherrole, role;
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'DENY MATCH {*} ON GRAPH secretDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
neo4j@system> GRANT MATCH {*} ON GRAPH publicDb, documentationDb NODES * TO role, otherrole;
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'GRANT MATCH {*} ON GRAPH publicDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
```

Author: Don't deny and grant work the same way? So if deny doesn't work on a role assigned to an auth rule, why would the grant work?
I think here, we want to show that instead of denying match on a specific databases, you can allow match on the dbs that you want.

Reviewer: If we have no auth rules involved the granting and denying of privileges behave the same (well one gives the ability to see things and the other blocks it but just assignment wise).

Auth rules are only allowed roles that doesn't have any DENY privileges -> neither adding deny privileges to a role that are granted to an auth rule nor granting a role with deny privileges to an auth rule should work. That is what these two error codes (22ND1 and 22ND2) is about. However regular GRANT privileges should be fine from my understanding 

So I'd say if we see the same error for the GRANT as the DENY that is a bug, I've asked @phil198 about it and he will take a look (since I could also reproduce it where I didn't expect to)

From earlier on this page:
> If an auth rule fails to evaluate, for example, because it depends on a claim that the user's auth token does not contain, Neo4j will default to not assigning that role to the user.
> 
> When a role contains denied privileges, not assigning the role actually removes the denies, which means the user ends up with more privileges than they should.
>
> To avoid this, roles cannot have both `DENY` privileges and be granted to an auth rule.

Reviewer: We have a PR up to fix the bug where specifically `GRANT ... ON GRAPH ...` commands incorrectly failed (had you tried `GRANT ACCESS ON DATABASE publicDb TO rolename` you would not have seen the error)

Author: Yep, that one works. But what do you think I could do with the solution? Should I leave it as it is, even though it currently fails?

Reviewer: The fix has been merged so it will make the 2026.07 cut. And yeah leave the example looking like we expect the behaviour to be sound like a good idea (even if that makes it 'fail' due to a bug in the meantime)

Reviewer 2: Thanks @Hunterness

== Example scenario

The example assumes that you have a database called `secretDb` and you want to deny access to all nodes in that database for a role called `role`.
You also have two roles, called `role` and `otherrole`, and an auth rule, called `authrule`, where `role` is granted to `authrule` and `otherrole` is not.

Try denying privileges to the role:

[source, cypher]
----
DENY MATCH {*} ON GRAPH secretDb NODES * TO otherrole, role;
----

The query returns a chain of errors where GQLSTATUS 22ND1 is the cause of GQLSTATUS xref:errors/gql-errors/42001.adoc[42001]:

.GQLSTATUS error chain
[source, error]
----
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'DENY MATCH {*} ON GRAPH secretDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
----

The error message contains a subset of the original query that caused the issue.
Since `otherrole` is not granted to an auth rule, it is not included.

== Possible solutions

Consider if it is possible to implement the security model without using denied privileges.

> For example, you can replace the generic deny from step 1 in the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:

Reviewer: do we need the "step 1" part here? isn't there only one part now
```suggestion
For example, you can replace the generic deny from the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:
```

Given that a role is used by an auth rule:
[source, cypher]
----
CYPHER 25 GRANT ROLE role TO AUTH RULE authrule;
----

Try denying privileges from the role:
[source, cypher]
----

> DENY MATCH {*} ON GRAPH secret-db NODES * TO otherrole, role;

Reviewer 3: I tried this query out and got an error:
````
42I06: Syntax error or access rule violation - invalid input
Invalid input '-', expected: a database name, ',', 'ELEMENT', 'ELEMENTS', 'FOR', 'NODE', 'NODES', 'RELATIONSHIP' or 'RELATIONSHIPS'. (line 1, column 31 (offset: 30))
"DENY MATCH {*} ON GRAPH secret-db NODES * TO otherrole, PUBLIC"
                               ^
Show less
42001: Syntax error or access rule violation - invalid syntax
````
Is this a valid database name in the query example? Or it doesn't matter?

Author: It seems that dashes are not allowed. https://neo4j.com/docs/operations-manual/current/database-administration/standard-databases/naming-databases/

Author: But I can't create a rule to test it. I get the same error on each example I run from https://neo4j.com/docs/operations-manual/current/authentication-authorization/attribute-based-access-control/#create-auth-rule:
```
neo4j@neo4j> CREATE AUTH RULE salesRule
                 SET CONDITION abac.oidc.user_attribute('department') = 'sales';
42I06: syntax error or access rule violation - invalid input. Invalid input 'RULE', expected: a graph pattern. (line 1, column 13 (offset: 12))
"CREATE AUTH RULE salesRule"
             ^
  42001: syntax error or access rule violation - invalid syntax
```

Author: Hmm, something is odd on my side. No matter what query I run, I get the same error code.

Author: Never mind. I managed to reproduce the first error. The only problem here is the hyphen in the database name, which I'll remove.

Reviewer: I think AUTH RULES are only on Cypher 25 and the invalid input on RULE seems like it's a Cypher 5 error perhaps?

But yeah, either removing the `-` or escaping it would make sense
```
DENY MATCH {*} ON GRAPH `secret-db` NODES * TO otherrole, role;
```

Author: @Hunterness, could you please review this one 22ND1 and the next one 22ND2? I tried to improve them.

For example, you can replace the generic grant from the <<example-scenario-22nd1, Example scenario>> with the following more fine-grained alternative:
[source, cypher]
----

> GRANT MATCH {*} ON GRAPH public-db, documentation-db NODES * TO role, otherrole;

Author: However, I get the same error here, apart from the hyphen. So, there might be something wrong with the solution.

```
neo4j@system> GRANT MATCH {*} ON GRAPH publicDb, documentationDb NODES * TO role, otherrole;
22ND1: data exception - operation not allowed for roles that are granted to an AUTH RULE. Invalid input: 'GRANT MATCH {*} ON GRAPH publicDb NODES * TO role' is not allowed for roles that are granted to an AUTH RULE.
  42001: syntax error or access rule violation - invalid syntax
```

Author: Never mind. I've figured it out.
